msi/GitProduct.wxs

```diff
@@ -1,76 +1,170 @@
-<?xml version='1.0'?>
-
-<?if $(sys.BUILDARCH)=x64 ?>
-  <?define SixtyFourBit=64 ?>
-<?else?>
-  <?define SixtyFourBit ?>
-<?endif?>
-
-<Wix xmlns='http://wixtoolset.org/schemas/v4/wxs'
-     xmlns:swid='http://wixtoolset.org/schemas/v4/wxs/tag'>
+<?xml version='1.0'?>
+
+<?if $(sys.BUILDARCH)=x64 ?>
+  <?define SixtyFourBit=64 ?>
+<?else?>
+  <?define SixtyFourBit ?>
+<?endif?>
+
+<Wix xmlns='http://wixtoolset.org/schemas/v4/wxs'
+     xmlns:swid='http://wixtoolset.org/schemas/v4/wxs/tag'>
   <Product Id='*' Name='Git for Windows' Language='1033' Manufacturer='The Git Development Community'
-           Version='$(var.Version)'
-           UpgradeCode='abbb3e1d-fea0-4b84-9585-b4dc13cbc1b4'>
-    <Package InstallScope='perMachine' Compressed='yes' Keywords='Git' />
+           Version='$(var.Version)'
+           UpgradeCode='abbb3e1d-fea0-4b84-9585-b4dc13cbc1b4'>
+    <Package InstallScope='perMachine' Compressed='yes' Keywords='Git' />
     <swid:Tag Regid='git-for-windows.github.io' />
-
-    <MajorUpgrade DowngradeErrorMessage='A newer version of [ProductName] is already installed.' />
-    <MediaTemplate CabinetTemplate='git{0}.cab' CompressionLevel='high' EmbedCab='yes' />
-
-    <!-- Add/Remove Programs Properties -->
-    <Property Id='ARPCONTACT' Value='https://github.com/git-for-windows/git/wiki/Contact' />
-    <Property Id='ARPURLINFOABOUT' Value='https://git-for-windows.github.io/' />
-    <Property Id='ARPNOMODIFY' Value='1' />
-    <Property Id='ARPPRODUCTICON' Value='git.ico' />
-    <SetProperty Id='ARPINSTALLLOCATION' Value='[INSTALLDIR]' After='CostFinalize' />
-
+
+    <MajorUpgrade DowngradeErrorMessage='A newer version of [ProductName] is already installed.' />
+    <MediaTemplate CabinetTemplate='git{0}.cab' CompressionLevel='high' EmbedCab='yes' />
+    <Icon Id='git.ico' SourceFile='git.ico' />
+
+    <!-- Add/Remove Programs Properties -->
+    <Property Id='ARPCONTACT' Value='https://github.com/git-for-windows/git/wiki/Contact' />
+    <Property Id='ARPURLINFOABOUT' Value='https://git-for-windows.github.io/' />
+    <Property Id='ARPNOMODIFY' Value='1' />
+    <Property Id='ARPPRODUCTICON' Value='git.ico' />
+    <SetProperty Id='ARPINSTALLLOCATION' Value='[INSTALLDIR]' After='CostFinalize' />
+
     <!-- Other MSI specific properties -->
-    <Property Id='MSIFASTINSTALL' Value='7' />
+    <Property Id='MSIFASTINSTALL' Value='7' />
+
+    <!-- Git for Windows MSI Properties -->
+    <!-- CLIENVIRONMENT Property implementing the "Remember Property" pattern -->
+    <Property Id="CLIENVIRONMENT" Value="Bash">
+      <RegistrySearch Id="CliEnvironmentSearch" Root="HKLM" Key="SOFTWARE\GitForWindows" Name="CliEnvironment" Type="raw" />
+    </Property>
+    <CustomAction Id="SaveCmdLineCLIENVIRONMENTValue" Property="CMDLINE_CLIENVIRONMENT" Value="[CLIENVIRONMENT]" Execute="firstSequence" />

```

Comment:
I didn't pay attention to the date of that blog article. It was really well-written and explained the problem very well. It was so well written that I logically understood it without having to think about it so much. And, so when I came to "implement" the pattern, I didn't even see the "similarities" with the `<SetProperty />` custom action. In fact, I thought you may have been implying there was some "trick" such that `<SetProperty />` didn't work as expected. But after your comment above, I see now what you're saying. It reduces the code _a lot_, and is much more readable and understandable.

It might not be a bad idea for you to revisit that blog post and update it to use `<SetProperty />` instead or create a quick new post and link to it from the original post for users of WiX v3 (I think v3??) or higher.